src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
                 }
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(TaskCompletionSource<bool> tcs = null)
             {
                 // Wait for response headers to be read.
-                (bool wait, bool emptyResponse) = TryEnsureHeaders();
-                if (wait)
+                bool emptyResponse;
+                bool wait;
+
+                do
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
                     (wait, emptyResponse) = TryEnsureHeaders();
-                    Debug.Assert(!wait);
+                    if (wait)
+                    {
+                        await GetWaiterTask().ConfigureAwait(false);
+                        (wait, emptyResponse) = TryEnsureHeaders();
+                        Debug.Assert(!wait);
+                    }
+
+                    if (Response.StatusCode  == HttpStatusCode.Continue)
+                    {
+                        OnResponse100Continue();
+                        tcs?.TrySetResult(true);
+                    }
+                }
+                while ((uint)Response.StatusCode < 200);
```

Comment:
I'm not following this logic... what happens if we receive, e.g., status code == 174... we loop forever?

Reply:
I think the logic here is basically correct. Any 1xx response is not a final response, we need to eat it and receive another set of response headers. So if we receive 174, eat it and wait for another set of response headers.